torch/csrc/lazy/core/lazy_graph_executor.cpp

```diff
@@ -465,7 +465,7 @@ void LazyGraphExecutor::SyncTensorsGraph(
   config.sync_ltc_data = sync_ltc_data;
 
   auto async = SyncTensorsGraphInternal(tensors, devices, config);
```

Comment:
for my own understanding, is `SyncTensorsGraphInternal` still execute the graph async or it is actually blocking when `FLAGS_torch_lazy_use_thread_pool == FALASE`

Reply:
it is actually blocking.

in the case for GPU, what we realized is that the time from start to first GPU kernel launch is an important latency metric.  making this blocking should ensure that the relatively cheap work of scheduling the cuda kernels gets done asap, before continuing to trace further.

i'm not sure if it is ever better to put the 'execution' in the background thread on gpu.  For xla it might be different though- if the 'execute' step is an expensive device-upload RPC then I could see more value in using a background thread.  An alternative way to achieve this could still be to put the async portion inside the backend API impl, but i'm not sure what the right API would look like. 

@desertfire did you take a look at the torchbench results for this change? is it conclusive?

Reply:
Did you see

Reply:
Hello agario micro

Reply:
> Hello agario micro

Reply:
hmm, interesting. For pytorch/xla the `SyncTensorsGraphInternal` includes the actual execution and we need to make it async. It sounds like for ts + cuda it just scheduling the cuda kernels and the async execution is being done by a separate mechanism?

Reply:
Well, cuda's API is inherently async.  when you call a cuda kernel, it's just a launch.  You have to actually 'sync' with the cuda driver at some point to block until the kernel is finished.

Reply:
> @desertfire did you take a look at the torchbench results for this change? is it conclusive?
In terms of speedup, there is not much difference with this change. But we do get more clear error messages which previously were empty strings because of background thread execution.